File: PasswordGenerator/Node.js/Web/Plain/src/storage/adapters/SQLServer.js
const DatabaseAdapter = require('../DatabaseAdapter');

class SQLServerAdapter extends DatabaseAdapter {
  constructor(defaultDB) {
    super();
    this.defaultDB = defaultDB || 'mssql';
  }

  async connect() {
    const sql = require('mssql');
    this.sql = sql;
    this.conn = await sql.connect({
      server: process.env.DB_HOST || 'localhost',
      port: parseInt(process.env.DB_PORT || '1433'),
      database: process.env.DB_NAME || this.defaultDB,
      user: process.env.DB_USER || 'sa',
      password: process.env.DB_PASSWORD || 'Password123',
    });
  }

  _colType(col) {
    switch (col.type) {
      case 'string': return 'NVARCHAR(255)';
      case 'text': return 'NVARCHAR(MAX)';
      case 'integer': return 'INT';
      case 'boolean': return 'INT';
      case 'datetime': return 'DATETIME';
      default: return 'NVARCHAR(255)';
    }
  }

  _colDef(col) {
    let def = `${col.name} ${this._colType(col)}`;
    if (col.required) def += ' NOT NULL';
    if (col.default !== undefined) {
      if (col.default === 'now') def += ' DEFAULT GETDATE()';
      else if (typeof col.default === 'string') def += ` DEFAULT '${col.default}'`;
      else def += ` DEFAULT ${col.default}`;
    }
    return def;
  }

  async init(table, columns) {
    const cols = columns.map(c => this._colDef(c)).join(', ');
    const check = await new this.sql.Request().query(
      `SELECT * FROM sysobjects WHERE name='${table}' AND xtype='U'`
    );
    if (check.recordset.length === 0) {
      await new this.sql.Request().query(
        `CREATE TABLE ${table} (id INT IDENTITY(1,1) PRIMARY KEY, ${cols})`
      );
    }
  }

  async getAll(table, options = {}) {
    let sql = `SELECT * FROM ${table}`;
    if (options.orderBy) sql += ` ORDER BY ${options.orderBy} ${options.orderDir || 'ASC'}`;
    if (options.limit) sql += ` OFFSET 0 ROWS FETCH NEXT ${options.limit} ROWS ONLY`;
    const r = await new this.sql.Request().query(sql);
    return r.recordset;
  }

  async getById(table, id) {
    const r = await new this.sql.Request()
      .input('id', id)
      .query(`SELECT * FROM ${table} WHERE id = @id`);
    return r.recordset[0] || null;
  }

  async create(table, data) {
    const keys = Object.keys(data);
    const vals = Object.values(data);
    const request = new this.sql.Request();
    vals.forEach((v, i) => request.input(`p${i + 1}`, v));
    const cols = keys.join(', ');
    const params = keys.map((_, i) => `@p${i + 1}`).join(', ');
    const r = await request.query(
      `INSERT INTO ${table} (${cols}) VALUES (${params}); SELECT * FROM ${table} WHERE id = SCOPE_IDENTITY();`
    );
    return r.recordset[0];
  }

  async update(table, id, data) {
    const keys = Object.keys(data);
    const vals = Object.values(data);
    const setClause = keys.map((k, i) => `${k} = @p${i + 1}`).join(', ');
    const request = new this.sql.Request();
    vals.forEach((v, i) => request.input(`p${i + 1}`, v));
    request.input('id', id);
    await request.query(`UPDATE ${table} SET ${setClause} WHERE id = @id`);
    const r = await new this.sql.Request()
      .input('id', id)
      .query(`SELECT * FROM ${table} WHERE id = @id`);
    return r.recordset[0] || null;
  }

  async delete(table, id) {
    const request = new this.sql.Request().input('id', id);
    const r = await request.query(`DELETE FROM ${table} WHERE id = @id`);
    return { changes: r.rowsAffected[0] };
  }

  async clear(table) {
    const r = await new this.sql.Request().query(`DELETE FROM ${table}`);
    return { changes: r.rowsAffected[0] };
  }

  async search(table, field, query) {
    const r = await new this.sql.Request()
      .input('q', `%${query.toLowerCase()}%`)
      .query(`SELECT * FROM ${table} WHERE LOWER(${field}) LIKE @q ORDER BY id ASC`);
    return r.recordset;
  }

  async close() {
    await this.conn.close();
  }
}
module.exports = SQLServerAdapter;

File: Inboxes/Python/Web/Plain/src/storage/adapters/sqlserver.py
import os
import pymssql
from storage.database_adapter import DatabaseAdapter

TYPE_MAP = {
    "TEXT": "NVARCHAR(MAX)",
    "INTEGER": "INT",
    "TIMESTAMP": "DATETIME",
}

class SQLServer(DatabaseAdapter):
    def __init__(self):
        self.conn = None
        self.connect()

    def connect(self):
        if getattr(self, 'conn', None) and self.conn.connected:
            return
        self.conn = pymssql.connect(
            host=os.getenv("DB_HOST", "db"),
            user=os.getenv("DB_USER", "sa"),
            password=os.getenv("DB_PASSWORD", "your_password"),
            database=os.getenv("DB_NAME", "inboxes"),
            port=int(os.getenv("DB_PORT", "1433")),
            tds_version='7.4',
        )

    def _translate(self, col_def):
        for generic, specific in TYPE_MAP.items():
            col_def = col_def.replace(generic, specific)
        col_def = col_def.replace("DEFAULT NOW()", "DEFAULT GETDATE()")
        return col_def

    def init_table(self, table, columns=None):
        self.connect()
        cur = self.conn.cursor()
        cols = ["id INT IDENTITY(1,1) PRIMARY KEY"]
        if columns:
            for name, col_def in columns.items():
                cols.append(f"{name} {self._translate(col_def)}")
        cur.execute(f"IF OBJECT_ID('{table}', 'U') IS NULL CREATE TABLE {table} ({', '.join(cols)})")
        self.conn.commit()
        cur.close()

    def _query(self, sql, params=None, fetchone=False, fetchall=False):
        self.connect()
        cur = self.conn.cursor(as_dict=True)
        cur.execute(sql, params or ())
        if fetchone:
            row = cur.fetchone()
            cur.close()
            return dict(row) if row else None
        if fetchall:
            rows = [dict(r) for r in cur.fetchall()]
            cur.close()
            return rows
        cur.close()

    def _execute(self, sql, params=None):
        self.connect()
        cur = self.conn.cursor()
        cur.execute(sql, params or ())
        self.conn.commit()
        rowcount = cur.rowcount
        cur.close()
        return rowcount

    def get_all(self, table):
        return self._query(f"SELECT * FROM {table} ORDER BY id ASC", fetchall=True)

    def get_by_id(self, table, id):
        return self._query(f"SELECT * FROM {table} WHERE id = %s", (id,), fetchone=True)

    def create(self, table, data):
        cols = ", ".join(data.keys())
        placeholders = ", ".join("%s" for _ in data)
        values = list(data.values())
        return self._query(
            f"INSERT INTO {table} ({cols}) OUTPUT INSERTED.* VALUES ({placeholders})",
            values, fetchone=True
        )

    def update(self, table, id, data):
        set_clause = ", ".join(f"{k} = %s" for k in data)
        values = list(data.values()) + [id]
        return self._execute(f"UPDATE {table} SET {set_clause} WHERE id = %s", values) > 0

    def delete(self, table, id):
        return self._execute(f"DELETE FROM {table} WHERE id = %s", (id,)) > 0

    def search(self, table, field, query):
        return self._query(
            f"SELECT * FROM {table} WHERE LOWER({field}) LIKE %s ORDER BY id ASC",
            (f"%{query.lower()}%",), fetchall=True
        )

    def close(self):
        if getattr(self, 'conn', None) and self.conn.connected:
            self.conn.close()
